File: src/main/java/uk/co/developmentanddinosaurs/gatling/lambda/javaapi/protocol/LambdaProtocolBuilder.java
package uk.co.developmentanddinosaurs.gatling.lambda.javaapi.protocol;

import io.gatling.core.protocol.Protocol;
import io.gatling.javaapi.core.ProtocolBuilder;
import software.amazon.awssdk.services.lambda.LambdaClient;

public class LambdaProtocolBuilder implements ProtocolBuilder {

  private final uk.co.developmentanddinosaurs.gatling.lambda.protocol.LambdaProtocolBuilder wrapped;

  public LambdaProtocolBuilder(
      uk.co.developmentanddinosaurs.gatling.lambda.protocol.LambdaProtocolBuilder wrapped) {
    this.wrapped = wrapped;
  }

  public LambdaProtocolBuilder client(LambdaClient client) {
    return new LambdaProtocolBuilder(wrapped.client(client));
  }

  @Override
  public Protocol protocol() {
    return wrapped.protocol();
  }
}

File: src/main/scala/uk/co/developmentanddinosaurs/gatling/lambda/protocol/LambdaProtocolBuilder.scala
package uk.co.developmentanddinosaurs.gatling.lambda.protocol

import com.softwaremill.quicklens.ModifyPimp
import io.gatling.core.config.GatlingConfiguration
import software.amazon.awssdk.services.lambda.LambdaClient

import scala.language.implicitConversions

object LambdaProtocolBuilder {

  implicit def toLambdaProtocol(
      builder: LambdaProtocolBuilder
  ): LambdaProtocol = builder.build

  def apply(configuration: GatlingConfiguration): LambdaProtocolBuilder =
    LambdaProtocolBuilder(LambdaProtocol(configuration))
}

final case class LambdaProtocolBuilder(protocol: LambdaProtocol) {
  def client(client: LambdaClient): LambdaProtocolBuilder =
    this.modify(_.protocol.lambdaClient).setTo(client)

  def build: LambdaProtocol = protocol
}
